### PR TITLE
Fixed PackageManager.resolveActivity deprecation.

### PR DESCRIPTION
**Pull request by n1snt**
## Purpose / Description
Fixes `resolveActivity` depreciation issue in https://github.com/ankidroid/Anki-Android/issues/12364#issuecomment-1264426702
I'm still a bit unsure about the fix I implemented in `AdaptionUtil`, it does not seem clean enough. I would love to hear if there is any better way to implement it.

## How Has This Been Tested?

- Tests
- Physical device
- API 33 Emulator

## Checklist

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented on your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by david-allison — CHANGES_REQUESTED** [review]
Please look into the `compat` package for how we handle deprecation.

You can see how this works: https://github.com/ankidroid/Anki-Android/pull/12422

**Comment by n1snt** [comment]
Hey, @david-allison can I please get a review on this?

**Review by lukstbit — CHANGES_REQUESTED** [review]
I've looked at the code and I'm not sure this is how we should move forward. The code changes the call site and would require additional changes when we would get to 33 being the minimum API level.
I would recommend instead:

1. Create an extension method in CompatHelper which fully replicates the newer non deprecated version of the method. This would be:

   `fun PackageManager.resolveActivity(intent: Intent, flags: ResolveInfoFlags): ResolveInfo?`

    which will simply delegate to `compat`

2. Create the new  method in `Compat`, also we can't use the newer class ResolveInfoFlags so we need to create our own implementation. This would be something like this:

    ```
    fun resolveActivity(packageManager: PackageManager, intent: Intent, flags: ResolveInfoFlags): ResolveInfo?

    data class ResolveInfoFlags(val value: Long) {
        companion object {
            fun of(value: Long) = ResolveInfoFlags(value = value)
        }
    }
    ```
    In `Compatv21` the implementation would call the resolveActivity version which receives an int from our custom ResolveInfoFlags:

   `packageManager.resolveActivity(intent, flags.value.toInt())`

    In `CompatV33` the newer method will be used:

    `packageManager.resolveActivity(intent, ResolveInfoFlags.of(flags.value))`

3. Use the previously declared method in CompatHelper instead of the deprecated ones. Make sure to make the imports look like the platform one. So we should have:
    
     `packageManager.resolveActivity(intent, ResolveInflFlags.of(0))`

      and not:

      ` packageManager.resolveActivity(intent, Compat.ResolveInfoFlags.of(0))`

      Note that there's one more usage in CustomTabsHelper that you hadn't addressed. You can check usages by using the `Find in files` view with `CTRL+SHIFT+F`(Linux) and searching for `.resolveActivity(` on a PackageManager instance.

With the changes above, when 33 will be the minimum API level we can simply delete the code above and only change some imports to use the newer version(of course assuming the we don't get other deprecations).

@david-allison @mikehardy Any opinions on what I said above?

**Comment by n1snt** [comment]
Hey, @lukstbit thanks a lot for this write-up. 👍
It made me understand how compact is implemented here.
I have reimplemented the resolveActivity compact stuff, please check if everything is correct.

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt [hunk 1]
```diff
@@ -92,6 +94,25 @@ interface Compat {
      */
     fun <T : Parcelable?> getParcelableExtra(intent: Intent, name: String, clazz: Class<T>): T?
 
+    /**
+     * Retrieve various kinds of information related to the application packages that are currently
+     * installed on the device.
+     * @param context
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:100]: As the _@ params_ are empty they should be deleted.

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt [hunk 2]
```diff
@@ -92,6 +94,25 @@ interface Compat {
...
+     * @param intent
+     * @param flags
+     * @return Returns a ResolveInfo object containing the final activity intent that was determined
+     * to be the best action. Returns null if no matching activity was found. If multiple matching
+     * activities are found and there is no default set, returns a ResolveInfo object containing
+     * something else, such as the activity resolver.
+     */
+    fun resolveActivity(packageManager: PackageManager, intent: Intent, flags: ResolveInfoFlags): ResolveInfo?
+
+    data class ResolveInfoFlags(val value: Long) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:110]: This class should also have documentation to explain it's role:

```
/**
 * Placeholder class(replicating the API 33 ResolveInfoFlags class) which is used to
 * provide a consistent API on all supported Android versions.
 */
```

File: AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt [hunk 3]
```diff
@@ -138,9 +140,8 @@ object AdaptionUtil {
     }
 
     // https://stackoverflow.com/questions/47610456/how-to-detect-miui-rom-programmatically-in-android
-    @Suppress("deprecation") // resolveActivity
     private fun isIntentResolved(ctx: Context, intent: Intent): Boolean {
-        return ctx.packageManager.resolveActivity(intent, PackageManager.MATCH_DEFAULT_ONLY) != null
+        return ctx.packageManager.resolveActivity(intent, Compat.ResolveInfoFlags.of(0)) != null
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt:144]: The ResolveInfoFlags class should be used without the Compat prefix. Then, when the minimum API level will be 33,  we could simply change the imports:

```
    private fun isIntentResolved(ctx: Context, intent: Intent): Boolean {
        return ctx.packageManager.resolveActivity(intent, ResolveInfoFlags.of(PackageManager.MATCH_DEFAULT_ONLY.toLong())) != null
    }
```
I think these are the imports needed:

```
import com.ichi2.compat.Compat.ResolveInfoFlags
import com.ichi2.compat.CompatHelper.Companion.resolveActivity
```

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt [hunk 4]
```diff
@@ -2408,7 +2410,7 @@ abstract class AbstractFlashcardViewer :
                     }
                 }
                 if (intent != null) {
-                    if (packageManager.resolveActivity(intent, 0) == null) {
+                    if (packageManager.resolveActivity(intent, Compat.ResolveInfoFlags(0)) == null) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:2413]: The same issue like I mentioned below that our `ResolveInfoFlags` should be used without the Compat identifier. Make sure you are using the resolveActivity method from CompatHelper(it should be in italic as it's an extension method).

**Review by lukstbit — CHANGES_REQUESTED** [review]
Thanks for taking time to work on this. I've left some notes on minor issues, plus I'm mentioning another usage of resolveActivity that wasn't handled:

https://github.com/ankidroid/Anki-Android/blob/2eeed6570ec37f89405d5e68ccef39d3bd49fccf/AnkiDroid/src/main/java/com/ichi2/compat/customtabs/CustomTabsHelper.kt#L60

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt (hunk 2 above)
Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:114]: Done

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt (hunk 1 above)
Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt]: Can you please check if this is fixed in the latest commits?

File: AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt (hunk 3 above)
Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt]: Done

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt (hunk 4 above)
Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt]: Done

**Comment by n1snt** [comment]
@lukstbit I have also replaced the usage of resolveActivity in CustomTabsHelper which you mentioned in your comment above. ✅

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt (hunk 1 above)
Comment (lukstbit) [reply AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt]: It isn't fixed.

File: AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt [hunk 5]
```diff
@@ -138,9 +140,8 @@ object AdaptionUtil {
     }
 
     // https://stackoverflow.com/questions/47610456/how-to-detect-miui-rom-programmatically-in-android
-    @Suppress("deprecation") // resolveActivity
     private fun isIntentResolved(ctx: Context, intent: Intent): Boolean {
-        return ctx.packageManager.resolveActivity(intent, PackageManager.MATCH_DEFAULT_ONLY) != null
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt:143]: This isn't correct as `PackageManager.MATCH_DEFAULT_ONLY` doesn't have a value of 0. So it should be `ResolveInfoFlags.of(PackageManager.MATCH_DEFAULT_ONLY.toLong())`.

**Review by lukstbit — CHANGES_REQUESTED** [review]
One more issue below to fix and we are all set.

File: AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt (hunk 5 above)
Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt:143]: Made this change 👍

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt [hunk 6]
```diff
@@ -92,6 +94,29 @@ interface Compat {
...
+     */
+    fun resolveActivity(packageManager: PackageManager, intent: Intent, flags: ResolveInfoFlags): ResolveInfo?
+
+    /**
+     * Placeholder class(replicating the API 33 ResolveInfoFlags class) which is used to
+     * provide a consistent API on all supported Android versions.
+     */
+    data class ResolveInfoFlags(val value: Long) {
+        companion object {
+            fun of(value: Long) = ResolveInfoFlags(value = value)
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:116]: This is only used once - the constructor is used the other two times.

**Review by david-allison — CHANGES_REQUESTED** [review]
This looks good.

change requested:
I'd like documentation on what `ResolveInfoFlags(0)` means, and see if we can add this as a method on the data class to reduce confusion

One optional nitpick

File: AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt (hunk 6 above)
Comment (lukstbit) [reply AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:116]: This is more than  a nitpick, it's a valid observation and one that I missed when I reviewed. The platform `ResolveInfoFlags` class doesn't have a public constructor and only uses the `.of(value)` method to be intialized.

@n1snt Usages of the constructor(like `ResolveInfoFlags(0)`) should be replaced with `ResolveInfoFlags.of(0)` to replicate the platform class.

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:116]: I'd remove this class (isn't useful at all, only the app heavier and slower to compile) and make the method take a Long flag argument restricted to the [same flags of ResolveInfoFlags with a @LongDef](https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/content/pm/PackageManager.java;l=753-773;drc=bbed027319b21ad822443aea9d29433d14f28003)

Comment (lukstbit) [reply AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:116]: @BrayanDSO Let's be honest, the class is not going to affect the build time at all. My reasoning was to do as much work as possible now so we can migrate to the platform version later, if we keep the flags version then I don't see the point of fixing these deprecations now, we are doing work now and will be doing work later, double the work no real benefit.

I would say that the issue with these api level 33 deprecations should be closed:

- we are doing work now for a very far future when API 33 will be the minimum. We, all the people in this thread,  almost 100% will not be there when it happens(the app might not even exist by then).
- there's a chance that switching to the new backend will eliminate the code containing them.
- these deprecations are not breaking, just makes the methods more type safe/ kotlin friendly
- it clutters the Compat  related classes which should be kept for real deprecations(like a complete difference between old/new way of doing something(not just one extra class parameter))

What do you think?

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:116]: > Let's be honest, the class is not going to affect the build time at all.

It shouldn't be relevant right now. But things like this pile up with time and then it becomes relevant.

> we are doing work now for a very far future when API 33 will be the minimum. We, all the people in this thread, almost 100% will not be there when it happens(the app might not even exist by then).

100% agreed with your points, especially this one. And I'm always in favor of doing less and avoiding doing temporary job.

I think that we could add an annotation for "accepted deprecations" that would do the same function of the supress annotation, but only to a curated list of methods, so it's easier to find what should be fixed and what shouldn't

Comment (n1snt) [reply AnkiDroid/src/main/java/com/ichi2/compat/Compat.kt:116]: Replaced with `ResolveInfoFlags.of(0)` ✅

**Review by david-allison — APPROVED** [review]

**Comment by david-allison** [comment]
Pending conflict resolution

**Comment by n1snt** [comment]
@david-allison Fixed merge conflict.

File: AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt [hunk 7]
```diff
@@ -2414,7 +2416,7 @@ abstract class AbstractFlashcardViewer :
                     }
                 }
                 if (intent != null) {
-                    if (packageManager.resolveActivity(intent, 0) == null) {
+                    if (packageManager.resolveActivity(intent, ResolveInfoFlags.of(0)) == null) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/anki/AbstractFlashcardViewer.kt:2419]: In #13110 we introduced a class `ResolveInfoFlagsCompat` which better models the new API method parameter. That's the one that should be used.

File: AnkiDroid/src/main/java/com/ichi2/compat/CompatV33.kt [hunk 8]
```diff
@@ -12,7 +12,7 @@
  *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
  *
  *  You should have received a copy of the GNU General Public License along with
- *  this program.  If not, see <http://www.gnu.org/licenses/>.
+ *  this program.  If not, see <http://www.gnu.org/licenses/>.want to keep this issue as an 'easy first step' to onboard users, or to be handled on a case-by-case basis when it causes issues with PRs.
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/compat/CompatV33.kt:15]: Extra comment?!?!

File: AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt [hunk 9]
```diff
@@ -123,6 +126,25 @@ object AdaptionUtil {
             ("Archytas".equals(Build.PRODUCT, ignoreCase = true) || "Archimedes".equals(Build.PRODUCT, ignoreCase = true))
     }
 
+    fun canUseContextMenu(): Boolean {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/utils/AdaptionUtil.kt:129]: I'm not sure what's up with this piece of code?

**Review by lukstbit — CHANGES_REQUESTED** [review]
Thanks for still working in this.
It would be better if you would rebase this PR on top of the latest changes so it's easier to review.

**Pull request closed without merging**